Test failures with current version of dependencies

Running the tests with current versions of numba, numpy, pytest-asyncio fails and/or times out. The following is an incomplete summary:


```
FAILED tests/template/test_com_template.py::test_com_default - nbclient.exceptions.CellTimeoutError: A cell timed out while it was being executed, after 600 seconds.
FAILED tests/template/test_sum_template.py::test_sum_default - nbclient.exceptions.CellTimeoutError: A cell timed out while it was being executed, after 600 seconds.
FAILED tests/template/test_com_template.py::test_com_rotflip - nbclient.exceptions.CellExecutionError: An error occurred while executing the following cell:
FAILED tests/template/test_sum_template.py::test_sum_roi - nbclient.exceptions.CellExecutionError: An error occurred while executing the following cell:
FAILED tests/template/test_sumsig_template.py::test_sum_sig_default - nbclient.exceptions.CellExecutionError: An error occurred while executing the following cell:
FAILED tests/template/test_disk_template.py::test_disk_default - nbclient.exceptions.CellTimeoutError: A cell timed out while it was being executed, after 600 seconds.
FAILED tests/template/test_fem_template.py::test_fem_analysis - nbclient.exceptions.CellExecutionError: An error occurred while executing the following cell:
FAILED tests/template/test_fft_template.py::test_sum_fft_default - nbclient.exceptions.CellTimeoutError: A cell timed out while it was being executed, after 600 seconds.
ERROR tests/server/test_browse.py::test_browse_localfs_fail - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_cancel.py::test_cancel_udf_job - RuntimeError: Event loop is closed
ERROR tests/server/test_cancel.py::test_cancel_udf_job - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_cluster_status.py::test_local_cluster_details - RuntimeError: Event loop is closed
ERROR tests/server/test_cluster_status.py::test_local_cluster_details - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_dataset.py::test_load_raw_success - RuntimeError: Event loop is closed
ERROR tests/server/test_dataset.py::test_load_raw_success - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_dataset.py::test_load_raw_with_backends - RuntimeError: Event loop is closed
ERROR tests/server/test_dataset.py::test_load_raw_with_backends - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_dataset.py::test_load_raw_fail - RuntimeError: Event loop is closed
ERROR tests/server/test_dataset.py::test_load_raw_fail - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_dataset.py::test_dataset_delete - RuntimeError: Event loop is closed
ERROR tests/server/test_dataset.py::test_dataset_delete - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_dataset.py::test_initial_state_after_reconnect - RuntimeError: Event loop is closed
ERROR tests/server/test_dataset.py::test_initial_state_after_reconnect - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_dataset.py::test_prime_cache - RuntimeError: Event loop is closed
ERROR tests/server/test_dataset.py::test_prime_cache - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_download.py::test_download_hdf5 - RuntimeError: Event loop is closed
ERROR tests/server/test_download.py::test_download_hdf5 - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_startup.py::test_get_config - RuntimeError: Event loop is closed
ERROR tests/server/test_startup.py::test_get_config - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_startup.py::test_conn_is_disconnected - RuntimeError: Event loop is closed
ERROR tests/server/test_startup.py::test_conn_is_disconnected - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_startup.py::test_conn_connect_local - RuntimeError: Event loop is closed
ERROR tests/server/test_startup.py::test_conn_connect_local - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_browse.py::test_browse_localfs - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_startup.py::test_cluster_connect_error - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_run_job_delete_ds - RuntimeError: Event loop is closed
ERROR tests/server/test_startup.py::test_cluster_connect_error - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_job.py::test_run_job_delete_ds - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_startup.py::test_initial_state_empty - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_cancel_unknown_job - RuntimeError: Event loop is closed
ERROR tests/server/test_startup.py::test_initial_state_empty - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_startup.py::test_initial_state_w_existing_ds - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_cancel_unknown_job - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_job.py::test_run_with_all_zeros_roi - RuntimeError: Event loop is closed
ERROR tests/server/test_startup.py::test_initial_state_w_existing_ds - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_job.py::test_run_with_all_zeros_roi - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_startup.py::test_initial_state_analyses - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_run_job_update_analysis_parameters - RuntimeError: Event loop is closed
ERROR tests/server/test_startup.py::test_initial_state_analyses - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_job.py::test_run_job_update_analysis_parameters - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_state.py::test_job_state_remove - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_analysis_removal - RuntimeError: Event loop is closed
ERROR tests/server/test_state.py::test_job_state_remove - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_job.py::test_analysis_removal - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_token_auth.py::test_missing_token[GET-/] - RuntimeError: Event loop is closed
ERROR tests/server/test_token_auth.py::test_missing_token[GET-/] - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_job.py::test_create_compound_analysis - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_create_compound_analysis - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_token_auth.py::test_missing_token[GET-/api/datasets/detect/] - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_rerun_com_analysis - RuntimeError: Event loop is closed
ERROR tests/server/test_token_auth.py::test_missing_token[GET-/api/datasets/detect/] - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_token_auth.py::test_missing_token[GET-/api/datasets/schema/] - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_rerun_com_analysis - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_job.py::test_com_guess_rpc - RuntimeError: Event loop is closed
ERROR tests/server/test_token_auth.py::test_missing_token[GET-/api/datasets/schema/] - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_token_auth.py::test_missing_token[POST-/] - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_com_guess_rpc - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_token_auth.py::test_missing_token[POST-/] - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_token_auth.py::test_missing_token[POST-/api/datasets/detect/] - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_com_guess_rpc_need_to_run_job - RuntimeError: Event loop is closed
ERROR tests/server/test_job.py::test_com_guess_rpc_need_to_run_job - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_token_auth.py::test_missing_token[POST-/api/datasets/detect/] - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_shutdown.py::test_shutdown - RuntimeError: Event loop is closed
ERROR tests/server/test_token_auth.py::test_missing_token[POST-/api/datasets/schema/] - RuntimeError: Event loop is closed
ERROR tests/server/test_shutdown.py::test_shutdown - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_token_auth.py::test_missing_token[POST-/api/datasets/schema/] - RuntimeError: Cannot close a running event loop
ERROR tests/server/test_startup.py::test_start_server - RuntimeError: Event loop is closed
ERROR tests/server/test_startup.py::test_start_server - RuntimeError: Cannot close a running event loop
```

I suspect this is the pytest-asyncio release from last week - I'll dig a bit deeper now.

--- a/conftest.py
+++ b/conftest.py
@@ -550,17 +550,28 @@ def fixup_event_loop():
     adjust_event_loop_policy()
 
 
-@pytest.hookimpl(trylast=True)
-def pytest_fixture_post_finalizer(fixturedef, request):
-    """Called after fixture teardown"""
-    if fixturedef.argname == "event_loop":
-        # Work around: pytest-asyncio sets an empty event loop policy here,
-        # which breaks on windows, where we have to supply a specific
-        # event loop policy. Until this is fixed in pytest-asyncio, manually re-set
-        # the event policy here.
-        # See also: https://github.com/pytest-dev/pytest-asyncio/pull/192
-        asyncio.set_event_loop_policy(None)
-        adjust_event_loop_policy()
+@pytest.fixture(scope="session")
+def event_loop():
+    """
+    We need to use a session-scoped event loop, otherwise we get
+    errors like `RuntimeError: Cannot close a running event loop`.
+    """
+    loop = asyncio.get_event_loop_policy().new_event_loop()
+    yield loop
+    print("event loop teardown")
+    loop.run_until_complete(loop.shutdown_asyncgens())
+    loop.run_until_complete(loop.shutdown_default_executor())
+    loop.close()
+
+    # Work around: pytest-asyncio sets an empty event loop policy here,
+    # which breaks on windows, where we have to supply a specific
+    # event loop policy. Until this is fixed in pytest-asyncio, manually re-set
+    # the event policy here.
+    # See also: https://github.com/pytest-dev/pytest-asyncio/pull/192
+    # (this is probably fixed in the current version of pytest-asyncio,
+    # but we can't use that one on Python 3.6, which we still support)
+    asyncio.set_event_loop_policy(None)
+    adjust_event_loop_policy()
 
 
 @pytest.fixture(autouse=True)

--- a/conftest.py
+++ b/conftest.py
@@ -560,7 +560,8 @@ def event_loop():
     yield loop
     print("event loop teardown")
     loop.run_until_complete(loop.shutdown_asyncgens())
-    loop.run_until_complete(loop.shutdown_default_executor())
+    if hasattr(loop, 'shutdown_default_executor'):
+        loop.run_until_complete(loop.shutdown_default_executor())
     loop.close()
 
     # Work around: pytest-asyncio sets an empty event loop policy here,

--- a/pytest.ini
+++ b/pytest.ini
@@ -13,3 +13,4 @@ log_format = %(asctime)s %(levelname)s %(message)s
 log_date_format = %Y-%m-%d %H:%M:%S
 env =
     D:NUMBA_NUM_THREADS = 4
+asyncio_mode = auto

--- a/test_requirements.txt
+++ b/test_requirements.txt
@@ -1,7 +1,6 @@
 pytest>=6,<7
 pytest-cov
-# See https://github.com/LiberTEM/LiberTEM/issues/1187
-pytest-asyncio>0.11,<0.17
+pytest-asyncio>0.11
 nest-asyncio
 pytest-xdist
 pytest-env

--- a/test_requirements.txt
+++ b/test_requirements.txt
@@ -1,4 +1,4 @@
-pytest>=6,<7
+pytest>=6
 pytest-cov
 pytest-asyncio>0.11
 nest-asyncio

--- a/tests/template/test_com_template.py
+++ b/tests/template/test_com_template.py
@@ -29,7 +29,7 @@ def test_com_default(hdf5_ds_2, tmpdir_factory, lt_ctx, local_cluster_url):
     notebook = notebook_generator(conn, dataset, analysis, save=True)
     notebook = io.StringIO(notebook.getvalue())
     nb = nbformat.read(notebook, as_version=4)
-    ep = ExecutePreprocessor(timeout=600)
+    ep = ExecutePreprocessor(timeout=30)
     ep.preprocess(nb, {"metadata": {"path": datadir}})
     channels = [
         "field",

